Doc/extending/newtypes_tutorial.rst

```diff
@@ -0,0 +1,865 @@
+
+Python provides a :c:func:`Py_VISIT` macro that automates calling visit
+functions.  With :c:func:`Py_VISIT`, we can minimize the amount of boilerplate
+in ``Noddy_traverse``::
+
+   static int
+   Noddy_traverse(NoddyObject *self, visitproc visit, void *arg)
+   {
+       Py_VISIT(self->first);
+       Py_VISIT(self->last);
+       return 0;
+   }
+
+.. note::
+   The :c:member:`~PyTypeObject.tp_traverse` implementation must name its
+   arguments exactly *visit* and *arg* in order to use :c:func:`Py_VISIT`.
+
+Second, we need to provide a method for clearing any subobjects that can
+participate in cycles::
+
+   static int
+   Noddy_clear(NoddyObject *self)
+   {
+       Py_CLEAR(self->first);
+       Py_CLEAR(self->last);
+       return 0;
+   }
+
+Notice the use of the :c:func:`Py_CLEAR` macro.  It is the recommended and safe
+way to clear data attributes of arbitrary types while decrementing
+their reference counts.  If you were to call :c:func:`Py_XDECREF` instead
+on the attribute before setting it to *NULL*, there is a possibility
+that the attribute's destructor would call back into code that reads the
+attribute again (*especially* if there is a reference cycle).
+
+.. note::
+   You could emulate :c:func:`Py_CLEAR` by writing::
+
+      PyObject *tmp;
+      tmp = self->first;
+      self->first = NULL;
+      Py_XDECREF(tmp);
+
+   Nevertheless, it is much easier and less error-prone to always
+   use :c:func:`Py_CLEAR` when deleting an attribute.  Don't
+   try to micro-optimize at the expense of robustness!
+
+The deallocator ``Noddy_dealloc`` may call arbitrary code when clearing
+attributes.  It means the circular GC can be triggered inside the function.
+Since the GC assumes reference count is not zero, we need to untrack the object
+from the GC by calling :c:func:`PyObject_GC_UnTrack` before clearing members.
+Here is our reimplemented deallocator using :c:func:`PyObject_GC_UnTrack`
+and ``Noddy_clear``::
+
+   static void
+   Noddy_dealloc(NoddyObject *self)
+   {
+       PyObject_GC_UnTrack(self);
+       Noddy_clear(self);
+       Py_TYPE(self)->tp_free((PyObject *) self);
+   }
+
+Finally, we add the :const:`Py_TPFLAGS_HAVE_GC` flag to the class flags::
+
+   .tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE | Py_TPFLAGS_HAVE_GC,
+
+That's pretty much it.  If we had written custom :c:member:`~PyTypeObject.tp_alloc` or
+:c:member:`~PyTypeObject.tp_free` slots, we'd need to modify them for cyclic
+garbage collection.  Most extensions will use the versions automatically provided.
+
+
+Subclassing other types
+=======================
+
+It is possible to create new extension types that are derived from existing
+types. It is easiest to inherit from the built in types, since an extension can
+easily use the :class:`PyTypeObject` it needs. It can be difficult to share
+these :class:`PyTypeObject` structures between extension modules.
+
+In this example we will create a :class:`Shoddy` type that inherits from the
```

Comment:
Seeing `Shoddy` as the example name makes me think "This is an example of a shoddy class implementation".

Perhaps it would be clearer if the examples were just `Example` and `SubExample` rather than `Noddy` and `Shoddy`?

Reply:
Alternatively, since "the `Example` example class" can read strangely, we could rename the examples based on what they demonstrate:

* `ExtMinimal` (replacing `Noddy1`)
* `ExtWithAttributes` (replacing `Noddy2`)
* `ExtWithSetters` (replacing `Noddy3`)
* `ExtWithGC` (replacing `Noddy4`)
* `ExtWithParent` (replacing `Shoddy`)

Where the `Ext` prefix is short for `Extension type`.

Reply:
To be honest, the names "Noddy" and "Shoddy" look like a joke by the original author(s). Overall the examples are silly, but it's hard to come with better ones (that still fit in the tutorial's structure).

I don't like the `ExtSomething` names either. We probably need something simple yet evocative...

At least `Shoddy` could be renamed `SubList`, I think.

Reply:
Yeah, I'm not especially fond of the `Ext...` names either.

I do like `Minimal` as the name for the first example, since it does strip class definition down to its essence (i.e. no state, no behaviour).

`PublicAttrs` would describe the second example, since it exposes the data attributes directly.

`GetSetAttrs` would describe the third example, since it hides the data attributes behind getter and setter functions.

`CyclicRefs` would describe the fourth example, since it handles the case where the object ends up referencing itself (directly or indirectly).

And then `SubList` for the list subclassing example.

Reply:
I find it nice that the current 4 examples keep the same object name, since each builds on top of the previous one. It feels like a distraction for the object name to change each time. Perhaps call it `Something`?

Reply:
Yeah, that problem didn't occur to me until after I'd posted my last set of suggestions.

`Minimal` doesn't work, since that isn't true by the last example. `Something` feels *too* vague.

Perhaps `Custom`? After all, the main reason to be doing this is because none of the existing builtin or extension types serve your needs, and the automatic code generators aren't serving your purposes either.

Reply:
I think we want to compare the various uses of the name:
* C file name: `custom.c`, `something.c`, `noddy.c`
* object instantiation: `Custom()`, `Something()`, `Noddy()`
* C object struct: `CustomObject`, `Something` (`SomethingObject`??), `NoddyObject`
* C type struct: `CustomType`, `SomethingType` (???), `NoddyType`

"Example" sounds almost right, except in the "example ``Example`` class" case. "Custom" feels ok too.

Reply:
"the `Custom` example class" and "the example `Custom` class" both read OK to me (whereas those were the phrases that put me off `Example`), and I think it sounds OK in "`custom` module" and "`Custom` instance" as well.

So if it sounds OK to you as well, I'd suggest going with that.

Reply:
I think `Noddy` and `Shoddy` are pretty good names for examples. Let not kill jokes in the documentation for a programming language named after the comic group.

Reply:
I don't know if those names are jokes (what does "Noddy" mean exactly)? Those names are a bit confusing.

Reply:
While `Noddy` and `Shoddy` are presumably meant just in fun, I found myself distracted by their actual meanings while reviewing Antoine's changes:

Noddy = this is pointless ("Yeah, it's a bit of a noddy change, but the boss insisted")
Shoddy = this is poorly implemented ("They did a shoddy job and I'm not happy with it")

For PEP 426, I used names like `ComfyChair` and `SoftCushions` in the examples, and picking some thematically appropriate arbitrary nouns from a Monty Python sketch may work here, too.

Reply:
I think it's good to avoid confusion with bizarre names so will probably change to the more neutral "Custom".